controllers/release/adapter_test.go

```diff
@@ -322,6 +330,192 @@ var _ = Describe("Release adapter", Ordered, func() {
 		})
 	})
 
+	When("EnsureFinalPipelineIsProcessed is called", func() {
+		var adapter *adapter
+
+		AfterEach(func() {
+			_ = adapter.client.Delete(ctx, adapter.release)
+		})
+
+		BeforeEach(func() {
+			adapter = createReleaseAndAdapter()
+			adapter.releaseServiceConfig = releaseServiceConfig
+		})
+
+		It("should do nothing if the Release final pipeline is complete", func() {
+			adapter.release.MarkFinalPipelineProcessing()
+			adapter.release.MarkFinalPipelineProcessed()
+
+			result, err := adapter.EnsureFinalPipelineIsProcessed()
+			Expect(!result.RequeueRequest && !result.CancelRequest).To(BeTrue())
+			Expect(err).NotTo(HaveOccurred())
+			Expect(adapter.release.IsFinalPipelineProcessing()).To(BeFalse())
+		})
+
+		It("should do nothing if the Release managed pipeline processing has not yet completed", func() {
+			adapter.release.MarkManagedPipelineProcessing()
+
+			result, err := adapter.EnsureFinalPipelineIsProcessed()
+			Expect(!result.RequeueRequest && !result.CancelRequest).To(BeTrue())
+			Expect(err).NotTo(HaveOccurred())
+			Expect(adapter.release.IsFinalPipelineProcessing()).To(BeFalse())
+		})
+
+		It("should do nothing if the Release tenant pipeline processing has not yet completed", func() {
+			adapter.release.MarkTenantPipelineProcessed()
+
+			result, err := adapter.EnsureFinalPipelineIsProcessed()
+			Expect(!result.RequeueRequest && !result.CancelRequest).To(BeTrue())
+			Expect(err).NotTo(HaveOccurred())
+			Expect(adapter.release.IsFinalPipelineProcessing()).To(BeFalse())
+		})
+
+		It("should requeue with error if fetching the Release final pipeline returns an error besides not found", func() {
+			adapter.ctx = toolkit.GetMockedContext(ctx, []toolkit.MockData{
+				{
+					ContextKey: loader.ReleasePipelineRunContextKey,
+					Err:        fmt.Errorf("some error"),
+				},
+			})
+			adapter.release.MarkManagedPipelineProcessingSkipped()
+
+			result, err := adapter.EnsureFinalPipelineIsProcessed()
+			Expect(result.RequeueRequest && !result.CancelRequest).To(BeTrue())
+			Expect(err).To(HaveOccurred())
+			Expect(adapter.release.IsFinalPipelineProcessing()).To(BeFalse())
+		})
+
+		It("should mark the Final Pipeline Processing as Skipped if the ReleasePlan isn't found", func() {
```

Comment:
I would expect this test case to be like this one
```
		It("should requeue the Release if any of the resources is not found", func() {
			adapter.ctx = toolkit.GetMockedContext(ctx, []toolkit.MockData{
				{
					ContextKey: loader.ReleasePlanContextKey,
					Err:        fmt.Errorf("Not Found"),
					Resource:   nil,
				},
			})

			result, err := adapter.EnsureTenantPipelineIsProcessed()
			Expect(result.RequeueRequest && !result.CancelRequest).To(BeTrue())
			Expect(err).To(HaveOccurred())
		})
```

Reply:
which you actually already have below, so this one can probably be deleted. The releasePlan has to be found... it shouldn't set it to skipped if it can't find it

Reply:
I have removed it